runtime/lua/vim/filetype.lua

```diff
@@ -1724,522 +1728,614 @@ local filename = {
+--     - Fast. Good rule of thumb is that it should be a short specific string (i.e. no quantifiers
+--       or character sets).
+--     - Specific. Good rules of thumb (from best to worst):
+--         - Full directory name (like '/etc/', '/log/').
+--         - Part of a rare enough directory name (like '/conf', 'git/').
+--         - Character sequence reasonably rarely used in real full paths (like 'nginx', 'cmus').
+--
+-- When modifying an existing regular pattern, make sure that it still fits its group.
+--
+--- @type table<string,vim.filetype.mapping>
 local pattern = {
   -- BEGIN PATTERN
-  ['.*/etc/a2ps/.*%.cfg'] = 'a2ps',
-  ['.*/etc/a2ps%.cfg'] = 'a2ps',
-  ['.*/usr/share/alsa/alsa%.conf'] = 'alsaconf',
-  ['.*/etc/asound%.conf'] = 'alsaconf',
-  ['.*/etc/apache2/sites%-.*/.*%.com'] = 'apache',
-  ['.*/etc/httpd/.*%.conf'] = 'apache',
-  ['.*/etc/apache2/.*%.conf.*'] = detect_apache,
-  ['.*/etc/apache2/conf%..*/.*'] = detect_apache,
-  ['.*/etc/apache2/mods%-.*/.*'] = detect_apache,
-  ['.*/etc/apache2/sites%-.*/.*'] = detect_apache,
-  ['access%.conf.*'] = detect_apache,
-  ['apache%.conf.*'] = detect_apache,
-  ['apache2%.conf.*'] = detect_apache,
-  ['httpd%.conf.*'] = detect_apache,
-  ['srm%.conf.*'] = detect_apache,
-  ['.*/etc/httpd/conf%..*/.*'] = detect_apache,
-  ['.*/etc/httpd/conf%.d/.*%.conf.*'] = detect_apache,
-  ['.*/etc/httpd/mods%-.*/.*'] = detect_apache,
-  ['.*/etc/httpd/sites%-.*/.*'] = detect_apache,
-  ['.*/etc/proftpd/.*%.conf.*'] = starsetf('apachestyle'),
-  ['.*/etc/proftpd/conf%..*/.*'] = starsetf('apachestyle'),
-  ['proftpd%.conf.*'] = starsetf('apachestyle'),
-  ['.*asterisk/.*%.conf.*'] = starsetf('asterisk'),
-  ['.*asterisk.*/.*voicemail%.conf.*'] = starsetf('asteriskvm'),
-  ['.*/%.aptitude/config'] = 'aptconf',
-  ['[mM]akefile%.am'] = 'automake',
-  ['.*/bind/db%..*'] = starsetf('bindzone'),
-  ['.*/named/db%..*'] = starsetf('bindzone'),
-  ['.*/build/conf/.*%.conf'] = 'bitbake',
-  ['.*/meta/conf/.*%.conf'] = 'bitbake',
-  ['.*/meta%-.*/conf/.*%.conf'] = 'bitbake',
-  ['.*%.blade%.php'] = 'blade',
-  ['bzr_log%..*'] = 'bzr',
-  ['.*enlightenment/.*%.cfg'] = 'c',
-  ['.*/%.cabal/config'] = 'cabalconfig',
-  ['.*/cabal/config'] = 'cabalconfig',
-  ['cabal%.project%..*'] = starsetf('cabalproject'),
-  ['.*/%.calendar/.*'] = starsetf('calendar'),
-  ['.*/share/calendar/.*/calendar%..*'] = starsetf('calendar'),
-  ['.*/share/calendar/calendar%..*'] = starsetf('calendar'),
-  ['sgml%.catalog.*'] = starsetf('catalog'),
-  ['.*/etc/defaults/cdrdao'] = 'cdrdaoconf',
-  ['.*/etc/cdrdao%.conf'] = 'cdrdaoconf',
-  ['.*/etc/default/cdrdao'] = 'cdrdaoconf',
-  ['.*hgrc'] = 'cfg',
-  ['.*%.[Cc][Ff][Gg]'] = {
-    detect.cfg,
-    -- Decrease priority to avoid conflicts with more specific patterns
-    -- such as '.*/etc/a2ps/.*%.cfg', '.*enlightenment/.*%.cfg', etc.
-    { priority = -1 },
+  ['/debian/'] = {
+    ['.*/debian/changelog'] = 'debchangelog',
```

Comment:
The that the outer and inner pattern are different format (anchored vs non-anchored is confusing). Either:

- Make all the outer patterns anchored: `/debian/` -> `.*/debian/.*`
- Make all the inner patterns non-anchored: `.*/debian/changelog` -> `/debian/changelog$`
   - This would also require a change to `vim.filetype.add` to add the `^$`.

Reply:
I prefer the second one as earlier preliminary testing showed them to be slightly faster. However, @clason expressed concern that it would make porting future filetype changes less straightforward.

If anything, I'd like it to be a separate, more visible in git logs PR.

Reply:
This was resolved in #29632.